File: Drivo.WebAPI/Services/InternalExamsService.cs
using Drivo.Entities;
using Drivo.Responses;
using Drivo.WebAPI.Repositories;

namespace Drivo.WebAPI.Services;

public class InternalExamsService
{
    public InternalExamsService(InternalExamsRepository internalExamsRepository)
    {
        InternalExamsRepository = internalExamsRepository;
    }

    private InternalExamsRepository InternalExamsRepository { get; }

    public async Task<List<InternalExamEntity>> GetInternalExamsByUserName(string userName)
    {
        return await InternalExamsRepository.GetInternalExamsByUserNameAsync(userName);
    }

    public async Task<InternalExamEntity> GetInternalExamByIdAsync(int internalExamId)
    {
        return await InternalExamsRepository.GetInternalExamByIdAsync(internalExamId);
    }

    public async Task<ActionResponse> AddInternalExamAsync(InternalExamEntity internalExam)
    {
        return await InternalExamsRepository.AddInternalExamAsync(internalExam);
    }

    public async Task<ActionResponse> UpdateInternalExamAsync(InternalExamEntity internalExam)
    {
        return await InternalExamsRepository.UpdateInternalExamAsync(internalExam);
    }

    public async Task<ActionResponse> RemoveInternalExamAsync(int internalExamId)
    {
        return await InternalExamsRepository.RemoveInternalExamAsync(await GetInternalExamByIdAsync(internalExamId));
    }
}

File: Drivo/src/app/services/internal-exams.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { firstValueFrom } from 'rxjs';
import { InternalExamEntity } from 'src/entities/InternalExamEntity';
import { environment } from 'src/environments/environment';
import { ActionResponse } from 'src/responses/action.response';

@Injectable({
  providedIn: 'root'
})
export class InternalExamsService {
  
  constructor(private httpClient: HttpClient) {}

  public async getInternalExamsAsync(): Promise<Array<InternalExamEntity>> {
    return await firstValueFrom(this.httpClient.get<Array<InternalExamEntity>>(`${environment.apiUri}/InternalExams`));
  }

  public async addInternalExamAsync(internalExam: InternalExamEntity): Promise<ActionResponse> {
    return await firstValueFrom(this.httpClient.post<ActionResponse>(`${environment.apiUri}/InternalExams`, internalExam));
  }

  public async updateInternalExamAsync(internalExam: InternalExamEntity): Promise<ActionResponse> {
    return await firstValueFrom(this.httpClient.put<ActionResponse>(`${environment.apiUri}/InternalExams`, internalExam));
  }

  public async removeInternalExamAsync(internalExamId: number): Promise<ActionResponse> {
    return await firstValueFrom(this.httpClient.delete<ActionResponse>(`${environment.apiUri}/InternalExams/${internalExamId}`));
  }

}
